docs/proposals/20190709-cluster-spec-crds.md

```diff
@@ -0,0 +1,325 @@
+    |                 |  |    is set               | |             |              |    
+    |                 |  |                         | |             |              |    
+    |                 |  | Set InfrastructureStage | |             |              |    
+    |                 |  | to "Provisioned"        | |             |              |    
+    |                 |<-+-------------------------| |             |              |    
+    |                 |  |                         | |             |              |    
+    |                 |  +-----------------------------------------+              |
+    |                 |                            +-+                            |
+    |                 |                             |                             |
+```
+Figure 1: Cluster provisioning process
+
+Figure 1 presents the sequence of actions involved in provisioning a cluster, highlighting the coordination required between the CAPI cluster controller and the provider infrastructure controller. 
+
+The creation of the Cluster and provider infrastructure objects are independent events. When a Cluster object is created, if the `InfrastructureRef` is not set, the cluster controller will set the `InfrastructureState` to "Pending". When a provider infrastructure object is created, the provider's controller will do nothing unless its owner reference is set to a cluster. 
+
+When the `InfrastructureRef` is set in the cluster, the cluster controller will retrieve the infrastructure object. If the object has not been seen before, it will start watching it. Also, if the object's owner is not set, it will set to the Cluster object and will set the cluster's `InfrastructureState` to "Provisioning". 
+
+When an infrastructure object is updated, the provider controller will check the owner reference. If it is set, it will retrieve the cluster object to obtain the required cluster specification and starts the provisioning process. When the process finishes, it sets the `InfrastructureStatusRef` in the cluster object. When the cluster controller detects the `InfrastructureStatusRef` is set, it sets the status to "Provisioned". 
+
+### States and Transitions
+
+The Cluster Controller has the responsibility of managing the state transitions during cluster lifecycle. In order to do so, it requires collaboration with the provider's controller to clearly signal the state transitions without requiring knowledge of the internals of the provider. 
+
+#### Pending
+
+The initial state when the Cluster object has been created but there is no provider's infrastructure associated with it. 
+
+##### Conditions
+- `Cluster.ProviderSpec.InsfrastructureRef` is an empty string
+
+#### Provisioning
+
+The cluster has a provider infrastructure object associated. The provider can start provisioning the infrastructure.
+
+##### Transition Conditions
+
+- The Cluster object has the `InfrastructureRef` field set 
```

Comment:
I would expect the InfrastructureRef to be set at creation time, I'm curious to know how this affects the Phases

Reply:
Assuming the reference is always set when the cluster is created will make the pending phase unnecessary.  

The current phases consider the possibility that the two objects, cluster and infrastructure are not created in a given order. It particularly opens the possibility of some tool watching for new clusters an creating the infrastructure for the cluster automatically.

Reply:
@vincepri I know there are some providers today that don't use the Cluster resource. If we can get them to switch to using it, primarily as a means to link related machines together, maybe we need to keep InfrastructureRef optional? Otherwise, they'd have to create a dummy reference to something...

Reply:
If it is too difficult to make the InfrastructureRef optional, maybe we can provide a Noop Infrastructure provider as part of `cluster-api`

Reply:
As I mentioned here https://github.com/kubernetes-sigs/cluster-api/pull/1137#discussion_r301756558, I'd prefer `InfrastructureRef` to be required like in Machines. If folks don't want to use the Cluster object, they can already do so today by not specifying a Cluster label on Machines.

Reply:
If you don't use a Cluster, you don't get machine.status.nodeRef, which means that machineSet.status is broken (the controller can't calculate ready/available/etc replicas). I'd like to suggest in a separate discussion that Cluster be required, but that's not up for debate as part of this proposal 😄.